src/transforms/gate/transform.rs

```diff
@@ -0,0 +1,268 @@
+
+use crate::{
+    conditions::Condition,
+    event::Event,
+    internal_events::GateEventsDropped,
+    transforms::{FunctionTransform, OutputBuffer},
+    transforms::gate::state::GateState,
+};
+
+#[derive(Clone)]
+pub struct Gate {
+    // Configuration parameters
+    pass_when: Option<Condition>,
+    open_when: Option<Condition>,
+    close_when: Option<Condition>,
+    max_events: usize,
+    auto_close: bool,
+    tail_events: usize,
+
+    // Internal variables
+    current_state: GateState,
+    buffer: VecDeque<Event>,
+    events_counter: usize,
+    events_dropped: Registered<GateEventsDropped>,
+    is_closing: bool,
+}
+
+impl Gate {
+    // This function is dead code when the feature flag `transforms-impl-gate` is specified but not
+    // `transforms-gate`.
+    #![allow(dead_code)]
+    pub fn new(
+        pass_when: Option<Condition>,
+        open_when: Option<Condition>,
+        close_when: Option<Condition>,
+        max_events: usize,
+        auto_close: bool,
+        tail_events: usize,
+    ) -> crate::Result<Self> {
+        let buffer = VecDeque::with_capacity(max_events);
+
+        Ok(Gate {
+            pass_when,
+            open_when,
+            close_when,
+            max_events,
+            auto_close,
+            tail_events,
+            events_dropped: register!(GateEventsDropped),
+            current_state: GateState::Closed,
+            buffer,
+            events_counter: 0,
+            is_closing: false,
+        })
+    }
+}
+
+impl FunctionTransform for Gate {
+    fn transform(&mut self, output: &mut OutputBuffer, event: Event) {
+        let (pass_gate, event) = match self.pass_when.as_ref() {
+            Some(condition) => {
+                let (result, event) = condition.check(event);
+                (result, event)
+            }
+            _ => (false, event)
+        };
+
+        let (open_gate, event) = match self.open_when.as_ref() {
+            Some(condition) => {
+                let (result, event) = condition.check(event);
+                (result, event)
+            }
+            _ => (false, event)
+        };
+
+        let (close_gate, event) = match self.close_when.as_ref() {
+            Some(condition) => {
+                let (result, event) = condition.check(event);
+                (result, event)
+            }
+            _ => (false, event)
+        };
+
+        if self.buffer.capacity() < self.max_events {
+            self.buffer.reserve(self.max_events);
+        }
+
+        if self.buffer.len() >= self.max_events {
+            self.buffer.pop_front();
+        }
+
+        self.buffer.push_back(event);
+
+        if pass_gate {
+            self.buffer.pop_back().map(|evt| output.push(evt));
+        } else if open_gate {
+            self.current_state = GateState::Open;
+            self.buffer.drain(..).for_each(|evt| output.push(evt));
+            self.events_counter = 0;
```

Comment:
I'd maybe set this in the `close_gate` block and call it `tail_events_counter` to make it clearer.